File: common/programmers/주차요금계산/주차요금계산_고진석.py
import math

def solution(fees, records):
    answer = []
    records = [record.split() for record in records]
    records = sorted(records, key = lambda x : x[1])

    res = dict()
    intime = -1
    carnum = ''
    for r in records:
        if r[2] == 'IN':
            if intime != -1:
                if carnum in res:
                    res[carnum] += totime("23:59") - intime
                else:
                    res[carnum] = totime("23:59") - intime
            intime = totime(r[0])
            carnum = r[1]
        else:
            if r[1] in res:
                res[r[1]] += totime(r[0]) - intime
            else:
                res[r[1]] = totime(r[0]) - intime
            intime = -1
            carnum = ''
    
    if intime != -1:
        if carnum in res:
            res[carnum] += totime("23:59") - intime
        else:
            res[carnum] = totime("23:59") - intime

    for time in res.values():
        if time <= fees[0]:
            answer.append(fees[1])
        else:
            answer.append(fees[1] + math.ceil((time - fees[0])/fees[2]) * fees[3])

    return answer

def totime(msg):
    h, m = map(int,msg.split(":"))
    return 60*h + m

print(solution(	[1, 461, 1, 10], ["00:00 1234 IN"]))

File: common/programmers/주차요금계산/주차요금계산_고진석.java
import java.util.*;

class 주차요금계산 {
    public int[] solution(int[] fees, String[] records) {
    	HashMap<String, Integer> map = new HashMap<>();
    	HashMap<String, Integer> res = new HashMap<>();
    	
    	for(String record : records) {
    		String[] tmp = record.split(" ");
    		if(tmp[2].equals("IN"))
    			map.put(tmp[1], toTime(tmp[0]));
    		else {
    			res.put(tmp[1], res.getOrDefault(tmp[1], 0) + toTime(tmp[0]) - map.get(tmp[1]));
    			map.remove(tmp[1]);
    		}
    	}
    	
    	for(String key : map.keySet()) {
    		res.put(key, res.getOrDefault(key, 0) + toTime("23:59") - map.get(key));
    	}
    	
    	int N = res.size();
        int[] answer = new int[N];
        
        Object[] keys = res.keySet().toArray();
        Arrays.sort(keys);
        
        for(int i = 0 ; i < N ; i++) {
        	int time = res.get(keys[i]);
    		answer[i] = fees[1];
        	if(time > fees[0]) {
        		time -= fees[0];
        		answer[i] += time % fees[2] == 0 ? fees[3] * time / fees[2] :  fees[3] * Math.ceil(time / fees[2]);
        	}
        }
        
        return answer;
    }
    
    public int toTime(String time) {
    	String tmp[] = time.split(":"); 
    	return Integer.parseInt(tmp[0]) * 60 + Integer.parseInt(tmp[1]);
    }
}